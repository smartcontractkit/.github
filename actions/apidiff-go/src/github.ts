import * as core from "@actions/core";
import * as github from "@actions/github";

const MARKDOWN_FINGERPRINT = "<!-- chainlink-apidiff-go -->";

export async function upsertPRComment(
  token: string,
  owner: string,
  repo: string,
  pull_number: number,
  commentBody: string,
): Promise<void> {
  const octokit = github.getOctokit(token);

  // 1. List all comments on the PR (issues API covers PR comments)
  const { data: comments } = await octokit.rest.issues.listComments({
    owner,
    repo,
    issue_number: pull_number,
    per_page: 100,
  });

  // 2. Look for a comment containing our identifier
  const existingComment = comments.find((c) =>
    c.body?.includes(MARKDOWN_FINGERPRINT),
  );
  const fingerprintedCommentBody = commentBody + `\n\n${MARKDOWN_FINGERPRINT}`;

  try {
    if (existingComment) {
      // 3a. Update the existing comment
      await octokit.rest.issues.updateComment({
        owner,
        repo,
        comment_id: existingComment.id,
        body: fingerprintedCommentBody,
      });
    } else {
      // 3b. Create a new comment
      await octokit.rest.issues.createComment({
        owner,
        repo,
        issue_number: pull_number,
        body: fingerprintedCommentBody,
      });
    }
  } catch (error) {
    core.warning(`Failed to upsert PR comment: ${error}`);
  }
}

export async function getSummaryUrl(
  token: string,
  owner: string,
  repo: string,
): Promise<string> {
  const runId = github.context.runId;

  const octokit = github.getOctokit(token);
  try {
    const { data } = await octokit.rest.actions.listJobsForWorkflowRun({
      owner,
      repo,
      run_id: runId,
    });

    if (data.jobs.length !== 1) {
      core.warning(
        `Expected exactly one job in workflow run, found ${data.jobs.length}. Cannot determine summary URL.`,
      );
      return "";
    }

    // Example format: https://github.com/smartcontractkit/chainlink-common/actions/runs/16783814681#summary-47529227742
    return `https://github.com/${owner}/${repo}/actions/runs/${runId}/#summary-${data.jobs[0].id}`;
  } catch (error) {
    core.warning(`Failed to get summary link: ${error}`);
    return "";
  }
}
