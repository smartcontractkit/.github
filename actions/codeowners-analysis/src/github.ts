import * as core from "@actions/core";
import * as github from "@actions/github";
import { GetResponseTypeFromEndpointMethod } from "@octokit/types";

const MARKDOWN_FINGERPRINT = "<!-- chainlink-codeowners-analysis -->";

export type OctokitType = ReturnType<typeof github.getOctokit>;
type ListFilesResponse = GetResponseTypeFromEndpointMethod<
  OctokitType["rest"]["pulls"]["listFiles"]
>;
export type PRFiles = ListFilesResponse["data"];

// docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#list-pull-requests-files
export async function getChangedFilesForPR(
  octokit: OctokitType,
  owner: string,
  repo: string,
  prNumber: number,
): Promise<PRFiles> {
  core.debug(`Comparing ${owner}/${repo} for PR ${prNumber}`);

  const prFiles = await octokit.paginate(octokit.rest.pulls.listFiles, {
    owner,
    repo,
    pull_number: prNumber,
    per_page: 100,
  });

  return prFiles;
}

export async function upsertPRComment(
  octokit: OctokitType,
  owner: string,
  repo: string,
  pull_number: number,
  commentBody: string,
): Promise<void> {
  // 1. List all comments on the PR (issues API covers PR comments)
  const { data: comments } = await octokit.rest.issues.listComments({
    owner,
    repo,
    issue_number: pull_number,
    per_page: 100,
  });

  // 2. Look for a comment containing our identifier
  const existingComment = comments.find((c) =>
    c.body?.includes(MARKDOWN_FINGERPRINT),
  );
  const fingerprintedCommentBody = commentBody + `\n\n${MARKDOWN_FINGERPRINT}`;

  try {
    if (existingComment) {
      // 3a. Update the existing comment
      await octokit.rest.issues.updateComment({
        owner,
        repo,
        comment_id: existingComment.id,
        body: fingerprintedCommentBody,
      });
    } else {
      // 3b. Create a new comment
      await octokit.rest.issues.createComment({
        owner,
        repo,
        issue_number: pull_number,
        body: fingerprintedCommentBody,
      });
    }
  } catch (error) {
    core.warning(`Failed to upsert PR comment: ${error}`);
  }
}

export async function getSummaryUrl(
  octokit: OctokitType,
  owner: string,
  repo: string,
): Promise<string> {
  const runId = github.context.runId;

  try {
    const { data } = await octokit.rest.actions.listJobsForWorkflowRun({
      owner,
      repo,
      run_id: runId,
    });

    if (data.jobs.length !== 1) {
      core.warning(
        `Expected exactly one job in workflow run, found ${data.jobs.length}. Cannot determine summary URL.`,
      );
      return "";
    }

    // Example format: https://github.com/smartcontractkit/chainlink-common/actions/runs/16783814681#summary-47529227742
    return `https://github.com/${owner}/${repo}/actions/runs/${runId}/#summary-${data.jobs[0].id}`;
  } catch (error) {
    core.warning(`Failed to get summary link: ${error}`);
    return "";
  }
}
