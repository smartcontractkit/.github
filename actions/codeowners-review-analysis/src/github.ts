import * as core from "@actions/core";
import * as github from "@actions/github";
import { GetResponseTypeFromEndpointMethod } from "@octokit/types";

const MARKDOWN_FINGERPRINT = "<!-- chainlink-codeowners-review-analysis -->";

export type OctokitType = ReturnType<typeof github.getOctokit>;
type ListFilesResponse = GetResponseTypeFromEndpointMethod<
  OctokitType["rest"]["pulls"]["listFiles"]
>;
export type PRFiles = ListFilesResponse["data"];

// docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#list-pull-requests-files
export async function getChangedFilesForPR(
  octokit: OctokitType,
  owner: string,
  repo: string,
  prNumber: number,
): Promise<PRFiles> {
  core.debug(`Comparing ${owner}/${repo} for PR ${prNumber}`);

  const prFiles = await octokit.paginate(octokit.rest.pulls.listFiles, {
    owner,
    repo,
    pull_number: prNumber,
    per_page: 100,
  });

  return prFiles;
}

export async function upsertPRComment(
  octokit: OctokitType,
  owner: string,
  repo: string,
  pull_number: number,
  commentBody: string,
): Promise<void> {
  // 1. List all comments on the PR (issues API covers PR comments)
  const { data: comments } = await octokit.rest.issues.listComments({
    owner,
    repo,
    issue_number: pull_number,
    per_page: 100,
  });

  // 2. Look for a comment containing our identifier
  const existingComment = comments.find((c) =>
    c.body?.includes(MARKDOWN_FINGERPRINT),
  );
  const fingerprintedCommentBody = commentBody + `\n\n${MARKDOWN_FINGERPRINT}`;

  try {
    if (existingComment) {
      // 3a. Update the existing comment
      await octokit.rest.issues.updateComment({
        owner,
        repo,
        comment_id: existingComment.id,
        body: fingerprintedCommentBody,
      });
    } else {
      // 3b. Create a new comment
      await octokit.rest.issues.createComment({
        owner,
        repo,
        issue_number: pull_number,
        body: fingerprintedCommentBody,
      });
    }
  } catch (error) {
    core.warning(`Failed to upsert PR comment: ${error}`);
  }
}

/**
 * Gets the CODEOWNERS file content from the repository.
 */
export async function getCodeownersFile(
  octokit: OctokitType,
  owner: string,
  repo: string,
) {
  const possibleFilenames = [
    ".github/CODEOWNERS",
    "CODEOWNERS",
    "docs/CODEOWNERS",
  ];

  const headSha = await getHeadSHAOfDefaultBranch(octokit, owner, repo);
  core.info(`Default branch HEAD SHA: ${headSha}`);

  for (const filename of possibleFilenames) {
    const response = await getFile(octokit, owner, repo, filename, headSha);
    if (response?.content) {
      return response;
    }
  }

  return undefined;
}

/**
 * Gets a file from specified ref.
 * See https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#get-repository-content
 */
async function getFile(
  octokit: OctokitType,
  owner: string,
  repo: string,
  path: string,
  ref: string,
) {
  try {
    core.info(`GitHub getContent: ${owner}/${repo} path:${path} ref:${ref}`);

    const { data } = await octokit.rest.repos.getContent({
      owner,
      repo,
      path,
      ref,
    });

    if (!("type" in data)) {
      throw new Error(
        `Expected a file at ${owner}/${repo}/${path}, but received no type information.`,
      );
    } else if (data.type !== "file") {
      // Not a file (could be dir, symlink, submodule)
      throw new Error(
        `Expected a file at ${owner}/${repo}/${path}, but received type "${data.type}".`,
      );
    } else if (!data.content) {
      throw new Error("No content found in getContent response.");
    }

    // Decode base64 content (GitHub may insert newlines in base64 payloads).
    const preprocessedContent = data.content.replace(/\n/g, "");
    const decodedContent = Buffer.from(preprocessedContent, "base64").toString(
      "utf-8",
    );

    const htmlUrl = `https://github.com/${owner}/${repo}/blob/${ref}/${encodeURI(path)}`;

    return { content: decodedContent, htmlUrl };
  } catch (error: any) {
    const requestPath = `${owner}/${repo}/${path}`;
    if (error?.status) {
      core.warning(
        `GitHub Request Error while getting file - ${requestPath}. (${error.status} - ${error.message})`,
      );
    } else {
      core.warning(
        `Unknown Error while getting file - ${requestPath} - ${String(error)}`,
      );
    }
    return { content: "", htmlUrl: "" };
  }
}

async function getHeadSHAOfDefaultBranch(
  octokit: OctokitType,
  owner: string,
  repo: string,
): Promise<string> {
  try {
    const response = await octokit.rest.repos.listCommits({
      owner,
      repo,
      per_page: 1,
    });
    if (response.data.length > 0) {
      return response.data[0].sha;
    }
    return "";
  } catch (error) {
    core.warning(`Failed to get default branch for ${owner}/${repo}: ${error}`);
    return "";
  }
}

export async function getSummaryUrl(
  octokit: OctokitType,
  owner: string,
  repo: string,
): Promise<string> {
  const runId = github.context.runId;

  try {
    const { data } = await octokit.rest.actions.listJobsForWorkflowRun({
      owner,
      repo,
      run_id: runId,
    });

    if (data.jobs.length !== 1) {
      core.warning(
        `Expected exactly one job in workflow run, found ${data.jobs.length}. Cannot determine summary URL.`,
      );
      return "";
    }

    // Example format: https://github.com/smartcontractkit/chainlink-common/actions/runs/16783814681#summary-47529227742
    return `https://github.com/${owner}/${repo}/actions/runs/${runId}/#summary-${data.jobs[0].id}`;
  } catch (error) {
    core.warning(`Failed to get summary link: ${error}`);
    return "";
  }
}
