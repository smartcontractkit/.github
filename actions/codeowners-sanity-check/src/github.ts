import * as core from "@actions/core";
import * as github from "@actions/github";

const MARKDOWN_FINGERPRINT = "<!-- chainlink-codeowners-sanity-check -->";

type Octokit = ReturnType<typeof github.getOctokit>;
export type CodeOwnersError = Awaited<
  ReturnType<Octokit["rest"]["repos"]["codeownersErrors"]>
>["data"]["errors"][0];

type CodeownersCheckResult =
  | { kind: "success"; errors: [] }
  | { kind: "errors"; errors: CodeOwnersError[] }
  | { kind: "not_found" }
  | { kind: "failure"; message: string };

export async function checkCodeOwners(
  token: string,
  owner: string,
  repo: string,
  ref: string,
): Promise<CodeownersCheckResult> {
  const octokit = github.getOctokit(token);

  try {
    const { data } = await octokit.rest.repos.codeownersErrors({
      owner,
      repo,
      ref,
    });

    if (data?.errors?.length > 0) {
      return { kind: "errors", errors: data.errors };
    }

    if (data?.errors?.length === 0) {
      return { kind: "success", errors: [] };
    }

    return { kind: "failure", message: "Unexpected response from Github" };
  } catch (error: any) {
    if (error?.status === 404) {
      return { kind: "not_found" };
    }

    return { kind: "failure", message: error.message };
  }
}

/**
 * Creates or updates a PR comment with the given body, identified by a fingerprint.
 */
export async function upsertPRComment(
  token: string,
  owner: string,
  repo: string,
  pull_number: number,
  commentBody: string,
): Promise<void> {
  core.debug("Upserting PR comment");
  const commentId = await findPRCommentByFingerprint(
    token,
    owner,
    repo,
    pull_number,
  );

  const octokit = github.getOctokit(token);
  const fingerprintedCommentBody = commentBody + `\n\n${MARKDOWN_FINGERPRINT}`;
  try {
    if (commentId === -1) {
      // not found
      core.debug("Creating new PR comment");
      await octokit.rest.issues.createComment({
        owner,
        repo,
        issue_number: pull_number,
        body: fingerprintedCommentBody,
      });
    } else {
      core.debug(`Updating existing PR comment ID: ${commentId}`);
      await octokit.rest.issues.updateComment({
        owner,
        repo,
        comment_id: commentId,
        body: fingerprintedCommentBody,
      });
    }
  } catch (error) {
    core.warning(`Failed to upsert PR comment: ${error}`);
  }
}

/**
 * Updates an existing PR comment with the given body, identified by a fingerprint.
 * If the comment does not exist, it will not create one.
 */
export async function updatePRComment(
  token: string,
  owner: string,
  repo: string,
  pull_number: number,
  commentBody: string,
): Promise<void> {
  core.debug("Updating PR comment");
  const commentId = await findPRCommentByFingerprint(
    token,
    owner,
    repo,
    pull_number,
  );

  if (commentId === -1) {
    core.info("No existing comment found, not updating.");
    return;
  }

  const octokit = github.getOctokit(token);
  const fingerprintedCommentBody = commentBody + `\n\n${MARKDOWN_FINGERPRINT}`;
  try {
    core.debug(`Updating existing PR comment ID: ${commentId}`);
    await octokit.rest.issues.updateComment({
      owner,
      repo,
      comment_id: commentId,
      body: fingerprintedCommentBody,
    });
  } catch (error) {
    core.warning(`Failed to update PR comment: ${error}`);
  }
}

async function findPRCommentByFingerprint(
  token: string,
  owner: string,
  repo: string,
  pull_number: number,
): Promise<number> {
  core.debug(`Finding PR comment for pull request #${pull_number}`);
  const octokit = github.getOctokit(token);
  const { data: comments } = await octokit.rest.issues.listComments({
    owner,
    repo,
    issue_number: pull_number,
    per_page: 100,
  });

  const existingComment = comments.find((c) =>
    c.body?.includes(MARKDOWN_FINGERPRINT),
  );
  const commentId = existingComment ? existingComment.id : -1;
  core.info(`Found existing comment ID: ${commentId}`);
  return commentId;
}

export async function getSummaryUrl(
  token: string,
  owner: string,
  repo: string,
): Promise<string> {
  const runId = github.context.runId;

  const octokit = github.getOctokit(token);
  try {
    const { data } = await octokit.rest.actions.listJobsForWorkflowRun({
      owner,
      repo,
      run_id: runId,
    });

    if (data.jobs.length !== 1) {
      core.warning(
        `Expected exactly one job in workflow run, found ${data.jobs.length}. Cannot determine summary URL.`,
      );
      return "";
    }

    // Example format: https://github.com/smartcontractkit/chainlink-common/actions/runs/16783814681#summary-47529227742
    return `https://github.com/${owner}/${repo}/actions/runs/${runId}/#summary-${data.jobs[0].id}`;
  } catch (error) {
    core.warning(`Failed to get summary link: ${error}`);
    return "";
  }
}
