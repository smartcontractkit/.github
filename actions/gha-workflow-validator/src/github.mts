import * as core from "@actions/core";
import * as github from "@actions/github";
import {
  GetResponseTypeFromEndpointMethod,
} from "@octokit/types";
import { join } from "node:path";
import { RequestError } from "@octokit/request-error";

export type Octokit = ReturnType<typeof github.getOctokit>;
type CompareResponse = GetResponseTypeFromEndpointMethod<Octokit["rest"]["repos"]["compareCommitsWithBasehead"]>;
export type GithubFiles = CompareResponse["data"]["files"];

export async function getComparison(
  octokit: Octokit,
  owner: string,
  repo: string,
  base: string,
  head: string,
): Promise<GithubFiles> {

  const diff = await octokit.rest.repos.compareCommitsWithBasehead({
    owner,
    repo,
    basehead: `${base}...${head}`
    // <before>...<after> or <earlier>...<later>
  });

  return diff.data.files;
}

/**
 * Gets the action definition file from Github and performs some error handling and retry logic.
 * If there's an error or no content is returned from github upon initial invocation, it will
 * retry the request once but switching between .yml and .yaml for the action definition
 * filename.
 * @param ghClient The octokit github client
 * @param requestObject The request object
 * @returns The action definition as a string or undefined if it failed
 */
export async function getActionFileFromGithub(
  ghClient: Octokit,
  owner: string,
  repo: string,
  repoPath: string,
  ref: string,
) {
  const actionFileGetter = async (path: string) => {
    try {
      const response = await ghClient.rest.repos.getContent({
        owner,
        repo,
        path,
        ref,
      });

      if ("content" in response.data) {
        return Buffer.from(response.data.content, "base64").toString();
      }
      throw Error("No content found in getContent response");
    } catch (error) {
      const requestPath = `${owner}/${repo}/${path}@${ref}`;
      if (error instanceof RequestError) {
        core.warning(
          `Encountered Github Request Error for ${requestPath}. (${error.status} - ${error.message})`
        );
      } else {
        core.warning(`Encountered Unknown Error for ${requestPath} - ${error}`);
      }
    }
  };


  const ymlPath = join(repoPath, "action.yml");
  const yamlPath = join(repoPath, "action.yaml");

  let actionFile = await actionFileGetter(ymlPath);
  if (!actionFile) {
    actionFile = await actionFileGetter(yamlPath);
  }
  return actionFile;
}

export async function commentOnPr(
  octokit: Octokit,
  owner: string,
  repo: string,
  prNumber: number,
  body: string,
) {
  await octokit.rest.issues.createComment({
    owner,
    repo,
    issue_number: prNumber,
    body,
  });
}