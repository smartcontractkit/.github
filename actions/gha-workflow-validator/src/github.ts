import * as core from "@actions/core";
import * as github from "@actions/github";
import { GetResponseTypeFromEndpointMethod } from "@octokit/types";
import { join } from "node:path";
import { COMMENT_HEADER } from "./strings.js";

export type Octokit = ReturnType<typeof github.getOctokit>;
type CompareResponse = GetResponseTypeFromEndpointMethod<
  Octokit["rest"]["repos"]["compareCommitsWithBasehead"]
>;
export type GithubFiles = CompareResponse["data"]["files"];

export async function getComparison(
  octokit: Octokit,
  owner: string,
  repo: string,
  base: string,
  head: string,
): Promise<GithubFiles> {
  core.debug(`Comparing ${owner}/${repo} commits ${base}...${head}`);

  const diff = await octokit.rest.repos.compareCommitsWithBasehead({
    owner,
    repo,
    basehead: `${base}...${head}`,
    // <before>...<after> or <earlier>...<later>
  });

  return diff.data.files;
}

/**
 * Gets the action definition file from Github and performs some error handling and retry logic.
 * If there's an error or no content is returned from github upon initial invocation, it will
 * retry the request once but switching between .yml and .yaml for the action definition
 * filename.
 * @param ghClient The octokit github client
 * @param requestObject The request object
 * @returns The action definition as a string or undefined if it failed
 */
export async function getActionFileFromGithub(
  octokit: Octokit,
  owner: string,
  repo: string,
  repoPath: string,
  ref: string,
) {
  const ymlPath = join(repoPath, "action.yml");
  const yamlPath = join(repoPath, "action.yaml");

  let actionFile = await getFileFromGithub(octokit, owner, repo, ymlPath, ref);
  if (!actionFile) {
    actionFile =  await getFileFromGithub(octokit, owner, repo, yamlPath, ref);
  }
  return actionFile;
}

export async function commentOnPrOrUpdateExisting(
  octokit: Octokit,
  owner: string,
  repo: string,
  prNumber: number,
  body: string,
) {
  const comments = await getAllCommentsOnPr(octokit, owner, repo, prNumber);
  const existingComment = comments.find(
    (comment) => comment.body?.startsWith(COMMENT_HEADER),
  );

  if (existingComment) {
    const response = await updateComment(octokit, owner, repo, existingComment.id, body);
    return { commentId: response.data.id, updatedAt: response.data.updated_at }
  } else {
    const response = await createComment(octokit, owner, repo, prNumber, body);
    return { commentId: response.data.id, createdAt: response.data.created_at };
  }
}

export async function deleteCommentOnPRIfExists(
  octokit: Octokit,
  owner: string,
  repo: string,
  prNumber: number,
) {
  const comments = await getAllCommentsOnPr(octokit, owner, repo, prNumber);
  const existingComment = comments.find(
    (comment) => comment.body?.startsWith(COMMENT_HEADER),
  );

  if (existingComment) {
    await octokit.rest.issues.deleteComment({
      owner,
      repo,
      comment_id: existingComment.id,
    });

    return true;
  }

  return false;
}

async function getAllCommentsOnPr(
  octokit: Octokit,
  owner: string,
  repo: string,
  prNumber: number,
) {
  core.debug(`Getting comments on PR ${prNumber}`);
  return await octokit.paginate(octokit.rest.issues.listComments, {
    owner: owner,
    repo: repo,
    issue_number: prNumber,
  });
}

async function createComment(
  octokit: Octokit,
  owner: string,
  repo: string,
  prNumber: number,
  body: string,
) {
  core.debug(`Commenting on PR ${prNumber}`);
  return octokit.rest.issues.createComment({
    owner,
    repo,
    issue_number: prNumber,
    body,
  });
}

async function updateComment(
  octokit: Octokit,
  owner: string,
  repo: string,
  commentId: number,
  body: string,
) {
  core.debug(`Updating comment ${commentId}`);
  return await octokit.rest.issues.updateComment({
    owner,
    repo,
    comment_id: commentId,
    body,
  });
}

async function getFileFromGithub(octokit: Octokit, owner: string, repo: string, path: string, ref: string) {
    try {
      core.debug(`Getting file through Github - ${owner}/${repo}/${path}@${ref}`);

      const response = await octokit.rest.repos.getContent({
        owner,
        repo,
        path,
        ref,
      });

      if ("content" in response.data) {
        return Buffer.from(response.data.content, "base64").toString();
      }
      throw Error("No content found in getContent response");
    } catch (error: any) {
      const requestPath = `${owner}/${repo}/${path}@${ref}`;
      if (error.status) {
        core.warning(
          `Encountered Github Request Error while getting file - ${requestPath}. (${error.status} - ${error.message})`,
        );
      } else {
        core.warning(`Encountered Unknown Error while getting file - ${requestPath} - ${error}`);
      }
    }
}