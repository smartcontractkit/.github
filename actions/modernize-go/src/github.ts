import * as core from "@actions/core";
import * as github from "@actions/github";

import { GetResponseTypeFromEndpointMethod } from "@octokit/types";

export type OctokitType = ReturnType<typeof github.getOctokit>;

export type PRReviewComment = NonNullable<
  NonNullable<
    Parameters<OctokitType["rest"]["pulls"]["createReview"]>[0]
  >["comments"]
>[0];

type ListFilesResponse = GetResponseTypeFromEndpointMethod<
  OctokitType["rest"]["pulls"]["listFiles"]
>;
export type PRFiles = ListFilesResponse["data"];

// docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#list-pull-requests-files
export async function getChangedFilesForPR(
  octokit: OctokitType,
  owner: string,
  repo: string,
  prNumber: number,
): Promise<PRFiles> {
  core.debug(`Comparing ${owner}/${repo} for PR ${prNumber}`);

  const prFiles = await octokit.paginate(octokit.rest.pulls.listFiles, {
    owner,
    repo,
    pull_number: prNumber,
    per_page: 100,
  });

  return prFiles;
}

/**
 * Posts one or more PR reviews containing inline suggestion comments.
 * Batches to avoid API limits.
 */
export async function postSuggestionReview(
  octokit: OctokitType,
  owner: string,
  repo: string,
  prNumber: number,
  headSha: string,
  comments: PRReviewComment[],
): Promise<void> {
  if (!comments.length) {
    core.info("No suggestion comments to post.");
    return;
  }

  core.startGroup(
    `Posting ${comments.length} suggestion comment(s) to PR #${prNumber}`,
  );
  for (let i = 0; i < comments.length; i += batchSize) {
    const chunk = comments.slice(i, i + batchSize);

    await octokit.rest.pulls.createReview({
      owner,
      repo,
      pull_number: prNumber,
      commit_id: headSha,
      event: "COMMENT",
      body: i === 0 ? reviewBody : undefined, // body only on first review (optional)
      comments: chunk.map((c) => ({
        path: c.path,
        body: c.body,
        // For multi-line replacements provide both start_line and line
        // For single-line replacements, GitHub accepts just `line`
        ...(c.start_line && c.start_line !== c.line
          ? {
              start_line: c.start_line,
              line: c.line,
              side: "RIGHT" as const,
              start_side: "RIGHT" as const,
            }
          : { line: c.line ?? c.start_line!, side: "RIGHT" as const }),
      })),
    });
  }
  core.endGroup();
}
