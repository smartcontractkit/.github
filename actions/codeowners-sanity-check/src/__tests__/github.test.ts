import { checkCodeOwners, upsertPRComment, updatePRComment } from "../github";
import { getNock, getTestOctokit } from "./__helpers__/test-utils.js";
import { vi, describe, it, expect } from "vitest";
import nock from "nock";

const nockBack = getNock();

vi.mock("@actions/core", async () => {
  return (await import("./__helpers__/test-utils.js")).coreLoggingStubs({});
});

describe(checkCodeOwners.name, () => {
  it("should check codeowners (valid)", async () => {
    const { nockDone } = await nockBack("dot-github-codeowners-valid.json");
    const octokit = getTestOctokit(nockBack.currentMode);
    const result = await checkCodeOwners(
      octokit,
      "smartcontractkit",
      ".github",
      "4d2f6bfd28367da87f493dd8854cbaa59f45a9b8",
    );
    expect(result).toEqual({ kind: "success", errors: [] });
    nockDone();
  });

  it("should check codeowners (invalid)", async () => {
    const { nockDone } = await nockBack(
      "chainlink-github-actions-codeowners-invalid.json",
    );
    const octokit = getTestOctokit(nockBack.currentMode);
    const result = await checkCodeOwners(
      octokit,
      "smartcontractkit",
      "chainlink-github-actions",
      "5a7543e418aa2dc57b75e85581eae30820a8c9ae",
    );
    expect(result).toMatchSnapshot();
    nockDone();
  });

  it("should check codeowners (missing)", async () => {
    const { nockDone } = await nockBack(
      "gha-org-workflows-codeowners-missing.json",
    );
    const octokit = getTestOctokit(nockBack.currentMode);
    const result = await checkCodeOwners(
      octokit,
      "smartcontractkit",
      "gha-org-workflows",
      "def9d574417d336ab11d157c340329dcabbf4366",
    );
    expect(result).toEqual({ kind: "not_found" });
    nockDone();
  });

  it("should fail gracefully on github server error", async () => {
    const ref = "def9d574417d336ab11d157c340329dcabbf4366";

    // mock response
    nock("https://api.github.com/")
      .get(
        `/repos/smartcontractkit/gha-org-workflows/codeowners/errors?ref=${ref}`,
      )
      .reply(500, {
        message: "simulated error",
      });

    const octokit = getTestOctokit(nockBack.currentMode);
    const result = await checkCodeOwners(
      octokit,
      "smartcontractkit",
      "gha-org-workflows",
      ref,
    );
    expect(result).toEqual({ kind: "failure", message: "simulated error" });
  });
});
