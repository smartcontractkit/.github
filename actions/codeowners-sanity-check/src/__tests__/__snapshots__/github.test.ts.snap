// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`checkCodeOwners > should check codeowners (invalid) 1`] = `
{
  "errors": [
    {
      "column": 31,
      "kind": "Unknown owner",
      "line": 2,
      "message": "Unknown owner on line 2: make sure @tateexon exists and has write access to the repository

  /chainlink-testing-framework/ @tateexon @kalverra @skudasov @anieeG @gheorghestrimtu
                                ^",
      "path": "CODEOWNERS",
      "source": "/chainlink-testing-framework/ @tateexon @kalverra @skudasov @anieeG @gheorghestrimtu
",
      "suggestion": "make sure @tateexon exists and has write access to the repository",
    },
    {
      "column": 61,
      "kind": "Unknown owner",
      "line": 2,
      "message": "Unknown owner on line 2: make sure @anieeG exists and has write access to the repository

  /chainlink-testing-framework/ @tateexon @kalverra @skudasov @anieeG @gheorghestrimtu
                                                              ^",
      "path": "CODEOWNERS",
      "source": "/chainlink-testing-framework/ @tateexon @kalverra @skudasov @anieeG @gheorghestrimtu
",
      "suggestion": "make sure @anieeG exists and has write access to the repository",
    },
    {
      "column": 69,
      "kind": "Unknown owner",
      "line": 2,
      "message": "Unknown owner on line 2: make sure @gheorghestrimtu exists and has write access to the repository

  /chainlink-testing-framework/ @tateexon @kalverra @skudasov @anieeG @gheorghestrimtu
                                                                      ^",
      "path": "CODEOWNERS",
      "source": "/chainlink-testing-framework/ @tateexon @kalverra @skudasov @anieeG @gheorghestrimtu
",
      "suggestion": "make sure @gheorghestrimtu exists and has write access to the repository",
    },
    {
      "column": 10,
      "kind": "Unknown owner",
      "line": 3,
      "message": "Unknown owner on line 3: make sure @tateexon exists and has write access to the repository

  /docker/ @tateexon @kalverra @skudasov @anieeG @gheorghestrimtu
           ^",
      "path": "CODEOWNERS",
      "source": "/docker/ @tateexon @kalverra @skudasov @anieeG @gheorghestrimtu
",
      "suggestion": "make sure @tateexon exists and has write access to the repository",
    },
    {
      "column": 40,
      "kind": "Unknown owner",
      "line": 3,
      "message": "Unknown owner on line 3: make sure @anieeG exists and has write access to the repository

  /docker/ @tateexon @kalverra @skudasov @anieeG @gheorghestrimtu
                                         ^",
      "path": "CODEOWNERS",
      "source": "/docker/ @tateexon @kalverra @skudasov @anieeG @gheorghestrimtu
",
      "suggestion": "make sure @anieeG exists and has write access to the repository",
    },
    {
      "column": 48,
      "kind": "Unknown owner",
      "line": 3,
      "message": "Unknown owner on line 3: make sure @gheorghestrimtu exists and has write access to the repository

  /docker/ @tateexon @kalverra @skudasov @anieeG @gheorghestrimtu
                                                 ^",
      "path": "CODEOWNERS",
      "source": "/docker/ @tateexon @kalverra @skudasov @anieeG @gheorghestrimtu
",
      "suggestion": "make sure @gheorghestrimtu exists and has write access to the repository",
    },
  ],
  "kind": "errors",
}
`;
